packages/gatsby-cli/src/reporter/redux/index.ts

```diff
@@ -46,6 +117,95 @@ export const dispatch = (action: ActionsUnion | Thunk): void => {
+        clearTimeout(stuckStatusDiagnosticTimer)
+      }
+
+      if (displayedStuckStatusDiagnosticWarning) {
+        // using nextTick here to prevent infinite recursion (report.warn would
+        // result in another call of this function and so on)
+        process.nextTick(() => {
+          const activitiesDiagnosticsMessage = generateStuckStatusDiagnosticMessage()
+          reporter.warn(
+            `This is just diagnostic information (enabled by GATSBY_DIAGNOSTIC_STUCK_STATUS_TIMEOUT):\n\nThere was activity since last diagnostic message. Log action:\n\n${JSON.stringify(
+              action,
+              null,
+              2
+            )}\n\nCurrently Gatsby is in: "${
+              store.getState().logs.status
+            }" state.${
+              activitiesDiagnosticsMessage.length > 0
+                ? `\n\nActivities preventing Gatsby from transitioning to idle state:\n\n${activitiesDiagnosticsMessage}`
+                : ``
+            }`
+          )
+        })
+        displayedStuckStatusDiagnosticWarning = false
+      }
+
+      if (currentStatus === ActivityStatuses.InProgress) {
+        stuckStatusDiagnosticTimer = setTimeout(
+          function logStuckStatusDiagnostic() {
+            displayingStuckStatusDiagnosticWarning = true
+            reporter.warn(
+              `This is just diagnostic information (enabled by GATSBY_DIAGNOSTIC_STUCK_STATUS_TIMEOUT):\n\nGatsby is in "${
+                store.getState().logs.status
+              }" state without any updates for ${(
+                (stuckStatusDiagnosticTimeoutDelay ?? 0) / 1000
+              ).toFixed(
+                3
+              )} seconds. Activities preventing Gatsby from transitioning to idle state:\n\n${generateStuckStatusDiagnosticMessage()}${
+                stuckStatusWatchdogTimeoutDelay
+                  ? `\n\nProcess will be terminated in ${(
+                      (stuckStatusWatchdogTimeoutDelay -
+                        stuckStatusDiagnosticTimeoutDelay) /
+                      1000
+                    ).toFixed(3)} seconds if nothing will change.`
+                  : ``
+              }`
+            )
+            displayingStuckStatusDiagnosticWarning = false
+            displayedStuckStatusDiagnosticWarning = true
+          },
+          stuckStatusDiagnosticTimeoutDelay
+        )
+      }
+    }
+
+    if (stuckStatusWatchdogTimeoutDelay) {
+      if (stuckStatusWatchdogTimer) {
+        clearTimeout(stuckStatusWatchdogTimer)
+      }
+
+      if (currentStatus === ActivityStatuses.InProgress) {
+        stuckStatusWatchdogTimer = setTimeout(
+          function fatalStuckStatusHandler() {
+            reporter.panic(
+              `Terminating the process (due to GATSBY_WATCHDOG_STUCK_STATUS_TIMEOUT):\n\nGatsby is in "${
+                store.getState().logs.status
+              }" state without any updates for ${(
+                (stuckStatusWatchdogTimeoutDelay ?? 0) / 1000
```

Comment:
If this was falsy (`null` or `0`) then it wouldn't pass the check on (new) line 185.
```suggestion
                stuckStatusWatchdogTimeoutDelay / 1000
```

Reply:
Done as part of https://github.com/gatsbyjs/gatsby/pull/26618/commits/e10a55644f3f0dd8a4bd419c7fa23075660e3e93